test/remote/gateways/remote_shift4_test.rb

```diff
@@ -0,0 +1,321 @@
+    assert response_result(response)['transaction']['invoice'].present?
+  end
+
+  def test_successful_purchase
+    auth_token = '8042677003331111'
+    response = @gateway.purchase(@amount, auth_token, @purchase_params)
+    assert success?(response)
+
+    assert_equal @options[:amount][:total], verify_amount(response)
+    assert successful_transaction?(response)
+  end
+
+  def test_failed_purchase
+    auth_token = '8038483489222221'
+    @purchase_params[:card][:expiration_date] = '085'
+    response = @gateway.purchase(@amount, auth_token, @purchase_params)
+
+    assert error?(response)
+    assert error_msg(response)
+  end
+
+  def test_failed_authorize
+    response = @gateway.authorize(@amount, @credit_card, @options.except(:access_token))
+
+    assert error?(response)
+    assert merchant_error(response)
+  end
+
+  def test_failed_capture
+    auth_token = '8038483489222221'
+    @capture_params[:card][:expiration_date] = '085'
+    response = @gateway.capture(@amount, auth_token, @capture_params)
+
+    assert error?(response)
+    assert error_msg(response)
+  end
+
+  def test_failed_refund
+    response = @gateway.refund(@amount, @credit_card, @refund_options.except(:access_token))
+
+    assert error?(response)
+    assert gen_error_msg(response)
+  end
+
+  def test_successful_refund
+    auth_token = '8042677003331111'
+    res = @gateway.purchase(@amount, auth_token, @options)
+    transaction = res.params['result'][0]['transaction']
+    card = res.params['result'][0]['card']
+    invoice = transaction['invoice'].to_s
+    auth_token = card['token']['value']
+    @refund_options[:invoice] = invoice
+
+    response = @gateway.refund(@amount, auth_token, @refund_options)
+
+    assert_equal @refund_options[:amount][:total], verify_amount(response)
+    assert_equal @refund_options[:invoice], response_result(response)['transaction']['invoice']
+  end
+
+  def test_successful_void
+    authorize_res = @gateway.authorize(@amount, @credit_card, @options)
+    invoice = authorize_res.params['result'][0]['transaction']['invoice']
+    @void_options[:invoice] = invoice
+    assert response = @gateway.void(@void_options)
+
+    assert_equal @refund_options[:amount][:total], verify_amount(response)
+    assert_equal @void_options[:invoice], response_result(response)['transaction']['invoice']
+  end
+
+  def test_failed_void
+    response = @gateway.void(@void_options.except(:access_token))
+    assert error?(response)
+    assert void_error_msg(response)
+    assert_equal error_code(response)['primaryCode'], 9803
+  end
+
+  private
+
+  def access_token_success_from(response)
+    credentials(response)['accessToken'].present?
+  end
+
+  def credentials(response)
+    response_result(response)['credential']
+  end
+
+  def successful_transaction?(response)
+    token(response).present?
+  end
+
+  def token(response)
+    response_result(response)['card']['token']['value']
+  end
+
+  def success?(response)
```

Comment:
`test_helper.rb` provides an `assert_success` and `assert_failure` method so I think this method and `def error?` can be removed and substituted with those helpers. For example: https://github.com/activemerchant/active_merchant/blob/89922760f0a9586c946ededc10ad1364ed52b87d/test/remote/gateways/remote_airwallex_test.rb#L19

Reply:
@drkjc I have tested them before creating mine, The existing methods and shift4 error response have different response structure I even tried to use but unable to use existing one.

Reply:
@ali-hassan The existing methods aren't working because of the changes [requested here](https://github.com/activemerchant/active_merchant/pull/4415/files#r890646169), and [here](https://github.com/activemerchant/active_merchant/pull/4415/files#r890646674). 

I've tested it out and once those are fixed the existing methods should work.